### PR TITLE
Encapsulate vehicle parts. Fill gaps when askew. Attempt two

### LINKED ISSUE #5684: Angled vehicles develop holes [$15]
While all this vehicle renaissance is going on, does it seem feasable to tackle that problem where cars at a 45-degree angle can be entered diagonally between otherwise solid parts? I think you know what I mean, how if two boards are adjacent vertically, at an angle you'll cross between them somehow. Ideally it would check the vehicle structure and make the space beyond functionally impassable/opaque from only that side. It'd probably be a tough fix, but it ought to happen at some point.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1548800-angled-vehicles-develop-holes?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mft-dev**
#### Summary
SUMMARY: Infrastructure "Encapsulate vehicle parts. Fill gaps when askew"
#### Purpose of change

Fixes #5684
Probably more? IDK

Encapsulate vehicle parts and allow for an implementation that solves the problem of gaps in vehicles at angles.

#### Describe the solution

Encapsulation of vehicle parts inside the vehicle class. Access to parts moves to `part(int)` for direct access to a part, or to the various `vehicle_part_range` implementations. The `parts` vector have been privatized and all callers have been changed to use either `part(int)` or an appropriate `vehicle_part_range` method

With encapsulation we can now choose what we show to outside code depending on needs. This means we can localize an implementation that widens a vehicle when it is askew. To support the solutions described in #5684 and #26739 a `fake_vehicle_mount` struct has been added.

In `vehicle::refresh()` we call `vehicle::refresh_fake_parts()` which then builds a map of edge parts (parts that don't have all four cardinal neighbors present). With this map we then build a map containing `fake_vehicle_mount` instances that fill the spots of the edge part's missing neighbors. We install a `vehicle_part` in the `parts` vector for each fake mount.

The `vehicle_part_range` ranges have been changed to take into account that the `parts` vector can contain fake parts, and most methods returning a range have gotten an overload to allow for including or excluding the fake parts as needed.

In `map::displace_vehicle()` we call `vehicle::update_active_fakes()` which then iterates over the fake mount map and activates/deactivates the part at the mount point depending on whether it is capable of filling an unintended gap.

Collisions to fake parts apply it's damage to the parent part immediately. 

#### Describe alternatives you've considered

Retrofitting path finding, LoS, lighting etc. to take adjacent diagonal impassables into account. It's not a viable solution at this moment

#### Testing

I've added some bare bones tests to `vehicle_tests.cpp` in the test project. The tests are tagged with `[vehicle]` and `[vehicle_fake]`.

I could not for the life of me get Visual Studio to detect an added file, no matter what hoops i jumped through, which is why they are placed in an existing file. Sorry :(

#### Additional context

There are some thorns to be aware of in this implementation. I ended up having to add more methods exposing access to the fake parts than I would like, namely `vehicle_interact.cpp` and `faction_camp.cpp` are... special... in how they deal with vehicles.

And it looks kind of ridiculous.

image

image

### CODE REVIEW

**Comment by mlangsdorf** [comment]
Would you please resubmit the vehicle::parts privatization as a separate PR?  It's a pretty straightforward set of commits, but it has a lot of impact on future development and is going to be the source of a lot of merge conflicts.  Get it out of the way as a separate PR, and then evaluating this PR is going to be easier.

File: tests/vehicle_test.cpp [hunk 1]
```diff
@@ -72,3 +75,209 @@ TEST_CASE( "add_item_to_broken_vehicle_part" )
...
+
+        WHEN( "The vehicle turns such that it is not perpendicular to a cardinal axis" ) {
+            veh->turn( 45 );
+            g->m.vehmove();
+            veh->idle( true );
+
+            THEN( "The gaps in the vehicle is filled by fake parts" ) {
+                int all_parts = 0;
+                for( const auto &p : veh->get_all_parts_incl_fake( true ) ) {
+                    if( p.part_index() >= 0 ) {
```
Comment (mlangsdorf) [inline tests/vehicle_test.cpp:134]: Travis CI is failing because this test is always true.

**Comment by Inglonias** [comment]
What.

File: src/vehicle.cpp [hunk 2]
```diff
@@ -2437,10 +2442,14 @@ item_location vehicle::part_base( int p )
 
 int vehicle::find_part( const item &it ) const
 {
-    auto idx = std::find_if( parts.begin(), parts.end(), [&it]( const vehicle_part & e ) {
-        return &e.base == &it;
-    } );
-    return idx != parts.end() ? std::distance( parts.begin(), idx ) : INT_MIN;
+    int index = INT_MIN;
```
[the remark this answers is not in the document]
Comment (mft-dev) [reply src/vehicle.cpp:2445]: > 
> 
> Would you please resubmit the vehicle::parts privatization as a separate PR? It's a pretty straightforward set of commits, but it has a lot of impact on future development and is going to be the source of a lot of merge conflicts. Get it out of the way as a separate PR, and then evaluating this PR is going to be easier.

Yes. I think that would make a lot more sense. The code is fairly robust since I've tested it extensively while implementing fake parts. And as you say, it's the code that is causing the merge conflicts. I will look at doing it later today - there is a rebase between the encapsulation parts and this PR in my branch, so will have to have a new look at veh_interact.cpp.

**Comment by mft-dev** [comment]
> Looks pretty good overall.  I need to go back and review your unit tests and the veh_interaction code.
> 
> MY biggest concern right now is though I can see how fake parts behave at the collision point, I don't see how fake parts interact with other damage sources (such as running over traps), being shot, or shock damage from collisions.  I may have just missed it.

You did not miss it, I failed to consider those sources. I'll give it a think once the encapsulation pr is up.

File: src/vehicle.cpp [hunk 3]
```diff
@@ -5596,15 +5627,195 @@ void vehicle::refresh()
...
+{
+    for( const std::pair<const point, fake_vehicle_mount> &fm : fake_mounts ) {
+        if( fm.second.fake_part_index == part_num ) {
+            return fm.second.active;
+        }
+    }
+    return !fake_part( part_num );
+}
+
+point vehicle::get_abs_diff( const point &one, const point &two ) const
```
[the remark this answers is not in the document]
Comment (mft-dev) [reply src/vehicle.cpp:5780]: > this should probably be in co-ordinates because it's not a vehicles specific thing.

I honestly thought such a method already existed, get_abs_diff is a monument to my failure to find it. I'll move it.

**Comment by EvgenijM86** [comment]
Why not use parts visual style to forbid movement? Like, if 2 parts have a shape like this:
```
|
|
```
then they represent vertical wall with no hole between them. We just need to store the information that parts were rotated, and when we have a scenario like this:
```
|s
Z|
```
zombie cannot move to s, if this vehicle was rotated.

This whole approach with creating fake parts seems like a very crude bandaid. If this is going to be merged in the game I would like to have an option to turn this off.

**Comment by EvgenijM86** [comment]
> > This whole approach with creating fake parts seems like a very crude bandaid. If this is going to be merged in the game I would like to have an option to turn this off.
> 
> Don't worry - there won't be an option.

I guess I will have to write a patch that adds an option to disable this.

File: src/vehicle_move.cpp [hunk 4]
```diff
@@ -564,9 +566,12 @@ veh_collision vehicle::part_collision( int part, const tripoint &p,
         }
     }
 
+    if( fake_part( ret.part ) ) {
+        ret.part = fake_mounts[parts[ret.part].mount].structural_part;
+    }
     // Damage armor before damaging any other parts
     // Actually target, not just damage - spiked plating will "hit back", for example
-    const int armor_part = part_with_feature( ret.part, VPFLAG_ARMOR, true );
+    int armor_part = part_with_feature( ret.part, VPFLAG_ARMOR, true );
```
Comment (BevapDin) [inline src/vehicle_move.cpp:574]: What's the reason for this change? You didn't change the rest of the function, so it seems to be working with the const just fine.

File: src/vpart_range.h [hunk 5]
```diff
@@ -101,14 +101,16 @@ class generic_vehicle_part_range
...
     public:
-        generic_vehicle_part_range( ::vehicle &v ) : vehicle_( v ) { }
+        generic_vehicle_part_range( ::vehicle &v, bool incl_fake = false ) : vehicle_( v ),
+            include_fake_( incl_fake ) { }
 
         // Templated because see top of file.
         template<typename T = ::vehicle>
         size_t part_count() const {
-            return static_cast<const T &>( vehicle_.get() ).parts.size();
+            return static_cast<const T &>( vehicle_.get() ).num_parts_incl_fake();
```
Comment (BevapDin) [inline src/vpart_range.h:113]: Shouldn't this depend on the value of `include_fake_`? When it's true, the fake parts are including in the count, otherwise they are excluded?

File: src/map.cpp [hunk 6]
```diff
@@ -544,19 +549,24 @@ vehicle *map::move_vehicle( vehicle &veh, const tripoint &dp, const tileray &fac
...
-            if( static_cast<size_t>( coll.part ) > veh.parts.size() ||
-                veh.parts[coll.part].removed ) {
+            if( !veh.valid_part( coll.part, true ) ||
+                veh.part( coll.part ).removed ) {
                 continue;
             }
+            int part_num = coll.part;
+            if( veh.fake_part( part_num ) ) {
+                part_num = veh.get_fake_parent( part_num );
+            }
```
Comment (BevapDin) [inline src/map.cpp:559]: These four lines are repeated several times. I suggest you move them into a function, so this can be shortened to
```suggestion
            const int part_num = get_non_fake_part( coll.part );
```
(Which either returns the input part index, or the fake parent.)

**Comment by ZhilkinSerg** [comment]
> > > This whole approach with creating fake parts seems like a very crude bandaid. If this is going to be merged in the game I would like to have an option to turn this off.
> > 
> > 
> > Don't worry - there won't be an option.
> 
> I guess I will have to write a patch that adds an option to disable this.

No, you don't have to do this.

**Comment by EvgenijM86** [comment]
> No, you don't have to do this.

I just don't want invisible walls appearing around my car, even if they are invisible. I want zombie be able to stand at point blank range near a car, and simply be unable to make diagonal step inside. As I understood Zombie will not be able to reach at point blank range near a car, because invisible wall will prevent them to stand there. If this is how it works I will investigate if it's possible to forbid diagonal step without creating invisible. Or at the very least make a patch that adds option to disable invisible walls, if it turns to be a really hard problem to just forbid diagonal step inside a car.

**Comment by anothersimulacrum** [comment]
>  I just don't want magical walls appearing around my car, even if they are invisible. I want zombie be able to stand at point blank range near a car, and simply be unable to make diagonal step inside. As I understood Zombie will not be able to reach at point blank range near a car, because invisible wall will prevent them to stand there. If this is how it works I will investigate if it's possible to forbid diagonal step without creating invisible walls. 

There's a reason this is the chosen solution for this issue - special casing diagonal movement not working for this case everywhere would be horrible.

**Comment by EvgenijM86** [comment]
And if those invisible walls actually collide with things while you drive then it would be horrible to make tight maneuvers around zombie, or even around complex terrain. Especially when you drive something like electric car that will easily get destroyed in collisions. It's like your car physically grows in size by ~1/3 of it's total squares when its rotated. Hopefully this is not the case and those invisible walls only prevent zombie to stand on those squares.

**Comment by ZhilkinSerg** [comment]
How about just stop making invalid assumptions and just test actual implementation? Not too controversial, isn't it?

**Comment by EvgenijM86** [comment]
> How about just stop making invalid assumptions and just test actual implementation? Not too controversial, isn't it?

Quote from pull request:
"Collisions to fake parts apply it's damage to the parent part immediately."

**Comment by ZhilkinSerg** [comment]
Yeah, nothing about "tight maneuvers" or anything like that.

File: src/vehicle.cpp [hunk 7]
```diff
@@ -2455,7 +2464,7 @@ item_group::ItemList vehicle_part::pieces_for_broken_part() const
 }
 
 std::vector<int> vehicle::parts_at_relative( const point &dp,
-        const bool use_cache ) const
+        const bool use_cache, bool include_fake ) const
 {
     if( !use_cache ) {
         std::vector<int> res;
```
Comment (BevapDin) [inline src/vehicle.cpp:2470]: This branch of the `if` doesn't check `include_fake`. It will always return a vector with fake parts, even when `include_fake` is false.

File: src/vehicle.cpp [hunk 8]
```diff
@@ -6845,11 +7056,88 @@ bounding_box vehicle::get_bounding_box()
...
+    return static_cast<int>( parts.size() ) - fake_part_count;
+}
+
+int vehicle::num_parts_incl_fake() const
+{
+    return static_cast<int>( parts.size() );
+}
+const vehicle_part &vehicle::cpart( int part_num ) const
+{
+    return const_cast<vehicle_part &>( parts[part_num] );
```
Comment (BevapDin) [inline src/vehicle.cpp:7090]:
```suggestion
    return parts[part_num];
```

File: src/vehicle.cpp [hunk 9]
```diff
@@ -6845,11 +7056,88 @@ bounding_box vehicle::get_bounding_box()
...
+    if( fake_part( part_num ) ) {
+        fake_vehicle_mount &found = find_fake( part_num );
+        return found.visible_part;
+    }
+    return -1;
+}
+
+vehicle_part &vehicle::part( int part_num )
+{
+    vehicle_part &vp = parts[part_num];
```
Comment (BevapDin) [inline src/vehicle.cpp:7115]:
```suggestion
    return parts[part_num];
```

File: src/vehicle.cpp [hunk 10]
```diff
@@ -6845,11 +7056,88 @@ bounding_box vehicle::get_bounding_box()
...
+}
+
+bool vehicle::valid_part( int part_num, bool include_fake ) const
+{
+    return part_num >= 0 && part_num < ( include_fake ?  num_parts_incl_fake() :  num_parts() );
+}
+
+bool vehicle::has_any_parts() const
+{
+    return static_cast<int>( parts.size() ) > 0;
```
Comment (BevapDin) [inline src/vehicle.cpp:7076]:
```suggestion
    return !parts.empty();
```

File: src/vehicle.cpp [hunk 11]
```diff
@@ -5596,15 +5627,195 @@ void vehicle::refresh()
...
+        if( fm.second.fake_part_index == part_num ) {
+            return fm.second.active;
+        }
+    }
+    return !fake_part( part_num );
+}
+
+point vehicle::get_abs_diff( const point &one, const point &two ) const
+{
+    return point( std::abs( one.x - two.x ), std::abs( one.y - two.y ) );
```
Comment (BevapDin) [inline src/vehicle.cpp:5782]:
```suggestion
    return ( one - two ).abs();
```
Taken from `manhatten_dist(...)` (defined in "line.cpp").

File: src/vehicle.cpp (hunk 8 above)
Comment (mft-dev) [reply src/vehicle.cpp:7090]: I honestly appreciate the effort. I had forgotten this PR was still open. Once I get some more free time I will have a look at your suggestions and create a PR with only the fake parts changes.

**Comment by mft-dev** [comment]
Sorry guys. Forgot that this PR was still open. Once I have some free time I will polish up the `fake parts` part of this PR. The encapsulation part was already merged. Closing this.
